src/mono/browser/runtime/jiterpreter-trace-generator.ts

```diff
@@ -3837,102 +3837,107 @@ function emit_simd_3 (builder: WasmBuilder, ip: MintOpcodePtr, index: SimdIntrin
+// implement shuffles on top of wasm's swizzle opcode by expanding the
+// element shuffle indices into byte indices
+function emit_shuffle (builder: WasmBuilder, ip: MintOpcodePtr, elementCount: number, checkBounds: boolean): boolean {
     const elementSize = 16 / elementCount,
         indicesOffset = getArgU16(ip, 3),
         constantIndices = get_known_constant_value(builder, indicesOffset);
-    mono_assert((elementSize === 2) || (elementSize === 4), "Unsupported shuffle element size");
 
     // Pre-load destination ptr
     builder.local("pLocals");
     // Load vec
     append_ldloc(builder, getArgU16(ip, 2), WasmOpcode.PREFIX_simd, WasmSimdOpcode.v128_load);
     if (typeof (constantIndices) === "object") {
-        // HACK: We have a known constant shuffle vector with char or int indices. Expand it to
+        // HACK: We have a known constant shuffle vector indices. Expand it to
         //  byte indices and then embed a new constant in the trace.
-        const newShuffleVector = new Uint8Array(sizeOfV128),
-            nativeIndices = (elementSize === 2)
-                ? new Uint16Array(constantIndices.buffer, constantIndices.byteOffset, elementCount)
-                : new Uint32Array(constantIndices.buffer, constantIndices.byteOffset, elementCount);
-        for (let i = 0, k = 0; i < elementCount; i++, k += elementSize) {
-            const elementIndex = nativeIndices[i];
-            for (let j = 0; j < elementSize; j++)
-                newShuffleVector[k + j] = (elementIndex * elementSize) + j;
+        let nativeIndices = new Uint8Array(constantIndices.buffer, constantIndices.byteOffset, elementCount);
+        if (elementCount !== 16) {
+            const newShuffleVector = new Uint8Array(sizeOfV128);
+            for (let i = 0, k = 0; i < elementCount; i++, k += elementSize) {
+                let elementIndex = 0;
+                for (let j = 0; j < elementSize; j++) {
+                    if (j == 0)
+                        elementIndex = nativeIndices[k + j];
+
+                    if (elementIndex >= elementCount || (j > 0 && nativeIndices[k + j] != 0)) {
+                        // this an invalid index make invalid in a safe way
+                        elementIndex = elementCount;
+                        break;
+                    }
+                }
+                for (let j = 0; j < elementSize; j++) {
+                    newShuffleVector[k + j] = (elementIndex * elementSize) + j;
+                }
+            }
+            nativeIndices = newShuffleVector;
         }
         // console.log(`shuffle w/element size ${elementSize} with constant indices ${nativeIndices} (${constantIndices}) -> byte indices ${newShuffleVector}`);
         builder.appendSimd(WasmSimdOpcode.v128_const);
-        builder.appendBytes(newShuffleVector);
+        builder.appendBytes(nativeIndices);
     } else {
         // Load indices (in chars)
```

Comment:
This is no longer chars (the comment was already wrong when you got here since it could have been int32s)